### PR TITLE
swan-accpy: Update swancustomenvironments

### PR DESCRIPTION
**Pull request by rodrigo-sobral**
Version: 0.0.11 -> 0.0.12

### CODE REVIEW

File: swan-accpy/Dockerfile [hunk 1]
```diff
@@ -30,6 +30,9 @@ RUN dnf install -y \
     dnf clean all && \
     rm -rf /var/cache/dnf
 
+# Install UV package manager for Python (faster that pip)
+RUN HOME=/home/${NB_USER} curl -LsSf https://astral.sh/uv/install.sh | sh
```
Comment (etejedor) [inline swan-accpy/Dockerfile:35]: This runs as user root, so root will be the owner of these files. And it seems you are creating something in the home directory of the user, who should own those files, not root.

File: swan-accpy/Dockerfile [hunk 2]
```diff
@@ -7,6 +7,7 @@ ARG NB_UID="1000"
 ARG BUILD_TAG=daily
 ENV VERSION_DOCKER_IMAGE=$BUILD_TAG
 ENV ACCPY_PATH="/opt/acc-py"
+ENV JAVA_HOME="/usr"
```
Comment (etejedor) [inline swan-accpy/Dockerfile:10]: Can you explain why we needed this again? The Spark connector does not find Java unless this is defined? Is it because it relies on JAVA_HOME and can't look for java in the path?

Also, why not defining JAVA_HOME to /usr/lib/jvm/java-11-openjdk-11.0.24.0.8-2.el9.x86_64?

File: swan-cern/scripts/before-notebook.d/10_spark.sh [hunk 3]
```diff
@@ -26,10 +26,9 @@ then
         '{load_extensions: $ARGS.named}' > $NBCONFIG/spark.json
 
   IPYTHON_KERNEL_CONFIG=$IPYTHONDIR/profile_default/ipython_kernel_config.py
-  sed -i "1s/^/c.InteractiveShellApp.extensions.append('sparkconnector.connector')\n/" \
-      $IPYTHON_KERNEL_CONFIG
-  sed -i "2s/^/c.InteractiveShellApp.extensions.append('sparkmonitor.kernelextension')\n/" \
-      $IPYTHON_KERNEL_CONFIG
+
+  echo -e "c.InteractiveShellApp.extensions.append('sparkmonitor.kernelextension')\n$( [ -f ${IPYTHON_KERNEL_CONFIG} ] && cat ${IPYTHON_KERNEL_CONFIG})" > ${IPYTHON_KERNEL_CONFIG}
```
Comment (etejedor) [inline swan-cern/scripts/before-notebook.d/10_spark.sh:30]: I would do these two lines in a separate commit. This is just an improvement to be able to insert at the beginning of a file when that file is empty. This is needed because when using custom environments, the IPYTHON_KERNEL_CONFIG file has not been created beforehand. All this should appear in the commit message.

File: swan/scripts/before-notebook.d/02_environment.sh [hunk 4]
```diff
@@ -130,6 +130,28 @@ if [ "$SOFTWARE_SOURCE" == "lcg" ]; then
   fi
 
   _log "Finished setting up CVMFS and user environment"
+else
+  # Configure NXCALS for customenvs
+  export SPARK_HOME=/cvmfs/sft.cern.ch/lcg/releases/spark/3.5.1-dfe72/x86_64-el9-gcc13-opt
+
+  # Configure necessary settings (it doesn't run without this)
```
Comment (etejedor) [inline swan/scripts/before-notebook.d/02_environment.sh:137]: This comment does not give any information. Why is the code below actually needed?

File: swan/scripts/before-notebook.d/02_environment.sh [hunk 5]
```diff
@@ -130,6 +130,28 @@ if [ "$SOFTWARE_SOURCE" == "lcg" ]; then
...
+  if [ ! -f $LOCAL_NB_NBEXTENSIONS ]; then 
+    echo "{
+      \"load_extensions\": {
+      }
+    }" > $LOCAL_NB_NBEXTENSIONS
+  fi
+  rm -f $NBCONFIG/notebook.json
+  ln -s $LOCAL_NB_NBEXTENSIONS $NBCONFIG/notebook.json
+
+  export KRB5CCNAME=$KRB5CCNAME_NB_TERM
```
Comment (etejedor) [inline swan/scripts/before-notebook.d/02_environment.sh:150]: This needs to be in a separate commit, with a message that explains that we need to set this variable here to point KRB5CCNAME to a writable directory. This is done for LCG already but not for custom envs, that's why we need it.

File: swan/scripts/before-notebook.d/02_environment.sh [hunk 6]
```diff
@@ -130,6 +130,28 @@ if [ "$SOFTWARE_SOURCE" == "lcg" ]; then
...
+  
+  # Configure NXCALs extensions
+  export SPARK_DIST_CLASSPATH=/cvmfs/sft.cern.ch/lcg/views/LCG_105a_nxcals_pro/x86_64-el9-gcc13-opt/nxcals/nxcals_java
+  export SPARK_CONF_DIR=/cvmfs/sft.cern.ch/lcg/etc/hadoop-confext/conf/etc/hadoop-nxcals/spark.hadoop-nxcals
```
Comment (etejedor) [inline swan/scripts/before-notebook.d/02_environment.sh:154]: I'm thinking whether to remove this for now. As I said in another PR, we can just start trying to make it work with spark connector and monitor plus pip install nxcals (manually) plus the minimal settings on top.

File: swan/scripts/before-notebook.d/02_environment.sh (hunk 4 above)
Comment (etejedor) [reply swan/scripts/before-notebook.d/02_environment.sh:137]: Actually, now that I read the code I don't get why it is needed. This is just loading extensions you previously loaded in other SWAN sessions, as per a file that got stored in your EOS user space. We should not need all this.

File: swan-accpy/Dockerfile (hunk 2 above)
Comment (rodrigo-sobral) [reply swan-accpy/Dockerfile:10]: Yes, JAVA_HOME needs to be explicitly defined in order to spark to find it. I tried to define JAVA_HOME using `readlink -f $(which java)` but isn't something that can be done in the Dockerfile. I avoided the path you mention for being too volatile, so I went for the safest option, which would be using '/usr', considering `/bin/java` (appended by the spark connector) would be pointing to the same java and it is more static/reliable

File: swan/scripts/before-notebook.d/02_environment.sh [hunk 7]
```diff
@@ -130,6 +130,12 @@ if [ "$SOFTWARE_SOURCE" == "lcg" ]; then
   fi
 
   _log "Finished setting up CVMFS and user environment"
+else
+  # Configure NXCALS for customenvs
+  export SPARK_HOME=/cvmfs/sft.cern.ch/lcg/releases/spark/3.5.1-dfe72/x86_64-el9-gcc13-opt
+
+  # This path needs to be set for sparkconnector_bundles getting properly linked (10_spark.sh)
+  mkdir -p $JPY_DIR/nbconfig
```
Comment (etejedor) [inline swan/scripts/before-notebook.d/02_environment.sh:138]: This commit can completely disappear.

File: swan-accpy/Dockerfile (hunk 2 above)
Comment (etejedor) [reply swan-accpy/Dockerfile:10]: I would remove this for now.

**Review by etejedor — APPROVED** [review]
